## Data Inspiration
<p>I was watching cricket match yesterday. India vs WestIndies. Both Rohit and Virat were well settled on crease with 200 run partnership and counting. virat was close to 150 and i saw a dialog box on TV stating kohli just needs another 154 runs to reach 10k Runs. To join the groups of elite Batsmen including Sachin, Pointing and Dravid. I was little surprised to see him reach close to the milestone so soon. </p>
<p>Then it stuck me, the amount of data in cricket. There's huge data in cricket. Data from 4's and 6's by a batsman in a match to number of balls faced in his test career.</p>
<p>It din't take me long to search for perfect website to find those stats. In India, cricket is a religion after all. I landed on cricinfo, searched for stats and found <a href='http://stats.espncricinfo.com/ci/content/records/335431.html'>this</a> webpage containing data of every single test match ever recorded or reported.</p>
<p>So, i dug in, scraped the data and analysed every stat i could imagine</p>

In [1]:
import requests

In [6]:
r = requests.get('http://stats.espncricinfo.com/ci/content/records/335431.html')
r.text[0:500]

'\n<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n<!-- hostname: web003, edition-view: espncricinfo-en-in, country: unknown, cluster: www, created: 2018-10-22 16:59:32 -->\n<html xmlns="http://www.w3.org/1999/xhtml" xmlns:fb="http://www.facebook.com/2008/fbml" xmlns:og="http://opengraphprotocol.org/schema/" xmlns:fb="http://developers.facebook.com/schema/" >\n<head>\n <script type="text/javascript">var _sf_startpt=(new Date())'

In [7]:
from bs4 import BeautifulSoup  
soup = BeautifulSoup(r.text, 'html.parser')

In [8]:
stats = soup.find_all('tr', attrs={'class':'data1'})
print("The resulting object is of type {}".format(type(stats)))
print("Total number of results, i.e total no of tests played till date are: {}".format(len(stats)))

The resulting object is of type <class 'bs4.element.ResultSet'>
Total number of results, i.e total no of tests played till date are: 738


The Whole Data is present in form of a list in the <i>bs4.element.ResultSet</i>. 
<p>Each test series is in form of an element in list.</p>
<p>Below is the structure of data in each element</p>

In [9]:
stats[0]

<tr class="data1" data-days="685654">
<td class="left"><a class="data-link" href="/ci/engine/series/60260.html">England in Australia Test Series</a></td>
<td class="left" nowrap="nowrap">1876/77</td>
<td nowrap="nowrap"></td>
<td class="left" nowrap="nowrap"><a class="data-link" href="/ci/content/team/1356.html">drawn</a></td>
<td nowrap="nowrap">1-1 (2)</td>
</tr>

### Data Structure of each table row of the data table
- The Number 685654 in data-days attribute. Such numbers notably unique are present in all elements. must mean something.
- The pathname of link to direct to page containing the test series results
- The Test Series teams playing. It gives information about the home team and the away team.
- The year the test series is played in
- A link inside 4th td tag leading to the winning team's page. It's throwing 404 (Page not Found) error for drawn series.
- The numerical Result of the series
- Total number of matches played in the series. ( Given in the Bracket )

So, the features (and it's values as per the first row) can be :
<p>
    <b>data-days</b> 685654<br>
    <b>Match link</b> /ci/engine/series/60260.html<br>
    <b>teams playing</b> England in Australia Test Series<br>
    <b>Year</b> 1876/77<br>
    <b>Team Won</b> Drawn<br>
    <b>Result</b> 1-1 (first number corresponds to winner)<br>
    <b>Matches Played</b> 2<br>
</p>

<p>The Dataframe would look something like this: </p>
<table>
    <thead>
    <tr>
        <td><b>data-days</b></td>
        <td><b>Series link</b></td>
        <td><b>Teams playing</b></td>
        <td><b>Year</b></td>
        <td><b>Team Won</b></td>
        <td><b>Result</b></td>
        <td><b>Matches Played</b></td>
    </tr>
    </thead>
    <tr>
        <td>685654<br></td>
        <td>/ci/engine/series/60260.html<br></td>
        <td>England in Australia Test Series<br></td>
        <td>1876/77<br></td>
        <td>Drawn<br></td>
        <td>1-1<br></td>
        <td>2<br></td>
    </tr>
</table>

- We can Further divide the Team playing into Home team and away team. Some of the Series have Series name embedded in this section. So, we can add another column <b>Series Name</b> to it. 
- We can divide the Year to series <b>starting Year</b> and <b>ending Year</b>
- The Result can be divided into <b>Home wins</b> and <b>away wins</b>

<p>With updated features, the dataframe would look like this: </p>
<table>
    <thead>
    <tr>
        <td><b>datadaysId</b></td>
        <td><b>Link</b></td>
        <td><b>Name</b></td>
        <td><b>Home</b></td>
        <td><b>Away</b></td>
        <td><b>startYear</b></td>
        <td><b>endYear</b></td>
        <td><b>TeamWon</b></td>
        <td><b>HomeWins</b></td>
        <td><b>AwayWins</b></td>
        <td><b>Matches</b></td>
    </tr>
    </thead>
    <tr>
        <td>685654</td>
        <td>/ci/engine/series/60260.html</td>
        <td></td>
        <td>Australia</td>
        <td>England</td>
        <td>1876</td>
        <td>1877</td>
        <td>Drawn</td>
        <td>1</td>
        <td>1</td>
        <td>2</td>
    </tr>
</table>

Now that we have format of the features, let's separate the values from the bs4 Resultset

In [10]:
final_result = stats[-2]
final_result

<tr class="data1" data-days="737346">
<td class="left"><a class="data-link" href="/ci/engine/series/1157748.html">West Indies in India Test Series</a></td>
<td class="left" nowrap="nowrap">2018/19</td>
<td nowrap="nowrap"></td>
<td class="left" nowrap="nowrap"><a class="data-link" href="/ci/content/team/6.html">India</a></td>
<td nowrap="nowrap">2-0 (2)</td>
</tr>

This series is the inspiration for my data. So, this data is updated till date.

In [ ]:
test_playing_nations = ['India', 'England', 'Australia', 'South Africa', 'West Indies', 'New Zealand', 'Pakistan', 'Sri Lanka', 'Zimbabwe', 'Bangladesh', 'Ireland', 'Afghanistan']
test_playing_nations

In [18]:
records = []
for series in stats:
    datadaysId = int(series['data-days'])
    Link = series.find('a').get('href')
    teams_text = series.find_all('td')[0].text
    years = series.find_all('td')[1].text
    winner = series.find_all('td')[3].text
    result = series.find_all('td')[4].text
    records.append((datadaysId, Link, teams_text, years, winner, result))

In [20]:
records[-3:]

[(737313,
  '/ci/engine/series/1119531.html',
  'Pataudi Trophy (India in England)',
  '2018',
  'England',
  '4-1 (5)'),
 (737346,
  '/ci/engine/series/1157748.html',
  'West Indies in India Test Series',
  '2018/19',
  'India',
  '2-0 (2)'),
 (737351,
  '/ci/engine/series/1157363.html',
  'Pakistan v Australia Test Series (in United Arab Emirates)',
  '2018/19',
  'Pakistan',
  '1-0 (2)')]

In [22]:
import pandas as pd
series = pd.DataFrame(records, columns=['datadaysID', 'link', 'teams_text', 'years', 'winner', 'result'])

In [24]:
series.head(10)

datadaysID                          link                        teams_text  \
0      685654  /ci/engine/series/60260.html  England in Australia Test Series   
1      686294  /ci/engine/series/60261.html   England in Australia Test Match   
2      686907  /ci/engine/series/60262.html   Australia in England Test Match   
3      687459  /ci/engine/series/60263.html  England in Australia Test Series   
4      687627  /ci/engine/series/60264.html   Australia in England Test Match   
5      687781  /ci/engine/series/60265.html  The Ashes (England in Australia)   
6      687803  /ci/engine/series/60266.html   England in Australia Test Match   
7      688342  /ci/engine/series/60267.html  The Ashes (Australia in England)   
8      688566  /ci/engine/series/60268.html  The Ashes (England in Australia)   
9      689073  /ci/engine/series/60269.html  The Ashes (Australia in England)   

     years     winner   result  
0  1876/77      drawn  1-1 (2)  
1  1878/79  Australia  1-0 (1)  
2     1880    England  1-0 (1)  
3  1881/82  Australia  2-0 (4)  
4     1882  Australia  1-0 (1)  
5  1882/83    England  2-1 (3)  
6  1882/83  Australia  1-0 (1)  
7     1884    England  1-0 (3)  
8  1884/85    England  3-2 (5)  
9     1886    England  3-0 (3)

In [25]:
series.tail(10)

datadaysID                            link  \
728      737152  /ci/engine/series/1115774.html   
729      737194  /ci/engine/series/1127280.html   
730      737213  /ci/engine/series/1119524.html   
731      737225  /ci/engine/series/1133250.html   
732      737236  /ci/engine/series/1135145.html   
733      737254  /ci/engine/series/1146713.html   
734      737263  /ci/engine/series/1142576.html   
735      737313  /ci/engine/series/1119531.html   
736      737346  /ci/engine/series/1157748.html   
737      737351  /ci/engine/series/1157363.html   

                                            teams_text    years       winner  \
728                 England in New Zealand Test Series  2017/18  New Zealand   
729                     Pakistan in Ireland Test Match     2018     Pakistan   
730                    Pakistan in England Test Series     2018        drawn   
731                    Afghanistan in India Test Match     2018        India   
732   Sobers/Tissera Trophy (Sri Lanka in West Indies)     2018        drawn   
733              Bangladesh in West Indies Test Series     2018  West Indies   
734              South Africa in Sri Lanka Test Series     2018    Sri Lanka   
735                  Pataudi Trophy (India in England)     2018      England   
736                   West Indies in India Test Series  2018/19        India   
737  Pakistan v Australia Test Series (in United Ar...  2018/19     Pakistan   

      result  
728  1-0 (2)  
729  1-0 (1)  
730  1-1 (2)  
731  1-0 (1)  
732  1-1 (3)  
733  2-0 (2)  
734  2-0 (2)  
735  4-1 (5)  
736  2-0 (2)  
737  1-0 (2)

In [26]:
series['winner'].unique()

array(['drawn', 'Australia', 'England', 'South Africa', 'West Indies',
       'India', 'Pakistan', 'New Zealand', 'Sri Lanka', 'Zimbabwe',
       'Bangladesh'], dtype=object)

In [27]:
len(series), series['datadaysID'].nunique()

(738, 729)

In [29]:
series['datadaysID'].value_counts()[:10]

731538    2
733630    2
732576    2
733764    2
698577    2
734895    2
731873    2
737152    2
736128    2
723789    1
Name: datadaysID, dtype: int64

In [30]:
series[series['datadaysID'] == 731538]

datadaysID                          link  \
475      731538  /ci/engine/series/60737.html   
476      731538  /ci/engine/series/60736.html   

                                teams_text    years        winner   result  
475       Pakistan in Zimbabwe Test Series  2002/03      Pakistan  2-0 (2)  
476  Sri Lanka in South Africa Test Series  2002/03  South Africa  2-0 (2)